Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,487 @@
+.. _a-conceptual-overview-of-asyncio:
+
+***************************************
+A Conceptual Overview of :mod:`asyncio`
+***************************************
+
+:Author: Alexander Nordin
+
+This article seeks to help you build a sturdy mental model of how ``asyncio``
+fundamentally works, helping you understand the how and why behind the
+recommended patterns.
+
+You might be curious about some key ``asyncio`` concepts.
+You'll be comfortably able to answer these questions by the end of this
+article.
+
+- What's happening behind the scenes when an object is ``await``\ ed?
+- How does ``asyncio`` differentiate between a task which doesn't need CPU-time
+  (such as a network request or file read) as opposed to a task that does
+  (such as computing n-factorial)?
+- How would I go about writing my own asynchronous variant of some operation?
+  Something like an async sleep, database request, and so on.
+
+--------------------------------------------
+A conceptual overview part 1: the high-level
+--------------------------------------------
+
+In part 1, we'll cover the main, high-level building blocks of ``asyncio``:
+the event loop, coroutine functions, coroutine objects, tasks and ``await``.
+
+==========
+Event Loop
+==========
+
+Everything in ``asyncio`` happens relative to the event loop.
+It's the star of the show.
+It's like an orchestra conductor.
+It's behind the scenes managing resources.
+Some power is explicitly granted to it, but a lot of its ability to get things
+done comes from the respect and cooperation of its teammates.
+
+In more technical terms, the event loop contains a queue of jobs (or "chunks
+of work") to be run.
+Some jobs are added directly by you, and some indirectly by ``asyncio``.
+The event loop pops a job from the queue and invokes it (or "gives it control"),
+similar to calling a function, then that job runs.
+Once it pauses or completes, it returns control to the event loop.
+The event loop will then move on to the next job in its queue and invoke it.
+This process repeats indefinitely.
+Even if the queue is empty, the event loop continues to cycle (somewhat
+aimlessly).
+
+Effective execution relies on tasks sharing well: a greedy job could hog
+control and leave the other tasks to starve, rendering the overall event loop
+approach rather useless.
+
+::
+
+   import asyncio
+
+   # This creates an event loop and indefinitely cycles through
+   # its queue of tasks.
+   event_loop = asyncio.new_event_loop()
+   event_loop.run_forever()
+
+=====================================
+Asynchronous Functions and Coroutines
+=====================================
+
+This is a regular 'ol Python function::
+
+   def hello_printer():
+       print(
+           "Hi, I am a lowly, simple printer, though I have all I "
```

Comment:
Can we stick to something simple like just hello world?

Reply:
I think the occasional bit of goofiness or fun makes the overall piece more engaging and offer a nice reprieve in the course of the read. In that light, what do you think?

Reply:
I believe HOWTO's (explainers) are meant to have room to be more casual and engaging rather than sterile and basic.